File: android-src/MainActivity.java
package com.smartstudy.admin;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import com.getcapacitor.BridgeActivity;
import com.capacitorjs.plugins.camera.CameraPlugin;

public class MainActivity extends BridgeActivity {

    private final BroadcastReceiver tokenReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context ctx, Intent intent) {
            String token = intent.getStringExtra("token");
            if (token == null) return;
            if (getBridge() != null)
                getBridge().triggerWindowJSEvent("fcmTokenRefresh",
                    "{\"token\":\"" + token + "\"}");
        }
    };

    @Override
    public void onCreate(Bundle savedInstanceState) {
        registerPlugin(CameraPlugin.class);
        registerPlugin(OcrPlugin.class);
        registerPlugin(BgSyncPlugin.class);
        registerPlugin(FcmTokenPlugin.class);
        registerPlugin(AdminPushPlugin.class);
        registerPlugin(GalleryPickerPlugin.class);
        super.onCreate(savedInstanceState);
        handleNavIntent(getIntent());
    }

    @Override
    public void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        handleNavIntent(intent);
    }

    @Override
    public void onResume() {
        super.onResume();
        IntentFilter filter = new IntentFilter("com.smartstudy.admin.FCM_TOKEN_REFRESH");
        registerReceiver(tokenReceiver, filter);
    }

    @Override
    public void onPause() {
        super.onPause();
        try { unregisterReceiver(tokenReceiver); } catch (Exception ignored) {}
    }

    private void handleNavIntent(Intent intent) {
        if (intent == null) return;
        String page = intent.getStringExtra("admin_nav_page");
        if (page == null || page.isEmpty()) return;
        new Handler(Looper.getMainLooper()).postDelayed(() -> {
            if (getBridge() != null)
                getBridge().triggerWindowJSEvent("adminNavTo",
                    "{\"page\":\"" + page + "\"}");
        }, 800);
    }

    @Override
    public void onBackPressed() {
        // দেখুন MainActivity.kt-এর একই মেথডের কমেন্ট — এই native double-back-exit
        // fallback (backPressedOnce/Toast) সরানো হলো, এখন সিদ্ধান্ত সম্পূর্ণভাবে
        // JS-এর (App.jsx handleBack) — single source of truth।
        if (getBridge() != null)
            getBridge().triggerWindowJSEvent("androidBackButton", "{}");
    }
}

File: android-src/MainActivity.kt
package com.smartstudy.admin

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.widget.Toast
import com.getcapacitor.BridgeActivity

class MainActivity : BridgeActivity() {
    private var backPressedOnce = false

    // FCM token refresh receiver
    private val tokenReceiver = object : BroadcastReceiver() {
        override fun onReceive(ctx: Context, intent: Intent) {
            val token = intent.getStringExtra("token") ?: return
            bridge?.triggerWindowJSEvent("fcmTokenRefresh", "{\"token\":\"$token\"}")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        registerPlugin(OcrPlugin::class.java)
        registerPlugin(BgSyncPlugin::class.java)
        registerPlugin(FcmTokenPlugin::class.java)
        super.onCreate(savedInstanceState)

        // FCM notification click — page navigate
        handleNavIntent(intent)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        handleNavIntent(intent)
    }

    override fun onResume() {
        super.onResume()
        val filter = IntentFilter("com.smartstudy.admin.FCM_TOKEN_REFRESH")
        registerReceiver(tokenReceiver, filter)
    }

    override fun onPause() {
        super.onPause()
        try { unregisterReceiver(tokenReceiver) } catch (_: Exception) {}
    }

    private fun handleNavIntent(intent: Intent?) {
        val page = intent?.getStringExtra("admin_nav_page") ?: return
        if (page.isBlank()) return
        // Bridge ready হওয়ার পরে JS এ event পাঠাও
        Handler(Looper.getMainLooper()).postDelayed({
            bridge?.triggerWindowJSEvent("adminNavTo", "{\"page\":\"$page\"}")
        }, 800)
    }

    override fun onBackPressed() {
        val handled = bridge?.triggerWindowJSEvent("androidBackButton", "{}") ?: false
        if (!handled) {
            if (backPressedOnce) { super.onBackPressed(); return }
            backPressedOnce = true
            Toast.makeText(this, "আবার Back চাপুন বন্ধ করতে", Toast.LENGTH_SHORT).show()
            Handler(Looper.getMainLooper()).postDelayed({ backPressedOnce = false }, 2000)
        }
    }
}
